osu.Game/Rulesets/Edit/Checks/CheckFewHitsounds.cs

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using System.Linq;
+using osu.Game.Audio;
+using osu.Game.Rulesets.Edit.Checks.Components;
+using osu.Game.Rulesets.Objects;
+
+namespace osu.Game.Rulesets.Edit.Checks
+{
+    public class CheckFewHitsounds : ICheck
+    {
+        /// <summary>
+        /// 2 measures (4/4) of 120 BPM, typically makes up a few patterns in the map.
+        /// This is almost always ok, but can still be useful for the mapper to make sure hitsounding coverage is good.
+        /// </summary>
+        private const int negligible_threshold_time = 4000;
+
+        /// <summary>
+        /// 4 measures (4/4) of 120 BPM, typically makes up a large portion of a section in the song.
+        /// This is ok if the section is a quiet intro, for example.
+        /// </summary>
+        private const int warning_threshold_time = 8000;
+
+        /// <summary>
+        /// 12 measures (4/4) of 120 BPM, typically makes up multiple sections in the song.
+        /// </summary>
+        private const int problem_threshold_time = 24000;
+
+        // Should pass at least this many objects without hitsounds to be considered an issue (should work for Easy diffs too).
+        private const int warning_threshold_objects = 4;
+        private const int problem_threshold_objects = 16;
+
+        private static readonly string[] hitsound_types = { HitSampleInfo.HIT_CLAP, HitSampleInfo.HIT_WHISTLE, HitSampleInfo.HIT_FINISH };
+
+        public CheckMetadata Metadata { get; } = new CheckMetadata(CheckCategory.Audio, "Few or no hitsounds");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateLongPeriodProblem(this),
+            new IssueTemplateLongPeriodWarning(this),
+            new IssueTemplateNoHitsounds(this)
+        };
+
+        private bool hasHitsounds;
+        private int objectsWithoutHitsounds;
+        private double lastHitsoundTime;
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            if (!context.Beatmap.HitObjects.Any())
+                yield break;
+
+            hasHitsounds = false;
+            objectsWithoutHitsounds = 0;
+            lastHitsoundTime = context.Beatmap.HitObjects.First().StartTime;
+
+            var hitObjectCount = context.Beatmap.HitObjects.Count;
+
+            for (int i = 0; i < hitObjectCount; ++i)
+            {
+                var hitObject = context.Beatmap.HitObjects[i];
```

Comment:
Side note but the way this algorithm is set up (iterating topmost objects in the outer loop and nested objects within the inner) is relying on two assumptions, namely:

* There are no overlapping objects / objects are sorted by their end time
* Nested hit objects are sorted by their end time, and their end time fits within the parent object's duration

Those are generally *fine* for the typical case but I'm afraid that it may break down a bit in cases where concurrent objects become a possibility (for osu! this isn't much of a problem, but for mania it sort of can become one). It may be better to collect all hitobjects and nested hitobjects to a list and then order them by end time explicitly to remove the aforementioned assumptions.

Reply:
That is a case I hadn't considered. After fixing it like you suggested my initial thought was to add a test, but thinking of a scenario in which this would be an issue is quite difficult without getting too specific. So am thinking I skip testing this case.

Reply:
Here's a test case that would fail with the old code but passes with the new.

```csharp
[Test]
public void TestConcurrentObjects()
{
    var hitObjects = new List<HitObject>();

    var notHitsounded = new List<HitSampleInfo> { new HitSampleInfo(HitSampleInfo.HIT_NORMAL) };
    var hitsounded = new List<HitSampleInfo>
    {
        new HitSampleInfo(HitSampleInfo.HIT_NORMAL),
        new HitSampleInfo(HitSampleInfo.HIT_FINISH)
    };

    var ticks = new List<HitObject>();
    for (int i = 1; i < 10; ++i)
        ticks.Add(new SliderTick { StartTime = 5000 * i, Samples = hitsounded });

    var nested = new MockNestableHitObject(ticks.ToList(), 0, 25000)
    {
        Samples = notHitsounded
    };
    nested.ApplyDefaults(new ControlPointInfo(), new BeatmapDifficulty());
    hitObjects.Add(nested);

    for (int i = 1; i < 10; ++i)
        hitObjects.Add(new HitCircle { StartTime = 10000 * i, Samples = notHitsounded });

    assertOk(hitObjects);
}
```

I'm not necessarily sure that this is permitted from a RC/mapping good practice standpoint, it's just a test case that would fail before due to not considering the possibility of nested HOs concurrent with normal HOs.

Reply:
I can't reproduce the failure. Reverting https://github.com/ppy/osu/pull/13669/commits/0c0fd291d9381d718668e0588d6fb5dee60e91c7 and testing this results in success. So does using the exact code from [the initial commit of `CheckFewHitsounds`](https://github.com/ppy/osu/blob/4b436b774dcb3c35d145410fb56edc12e2c66ebb/osu.Game/Rulesets/Edit/Checks/CheckFewHitsounds.cs).

In terms of the case being realistic/practical, it would be if we replace the slider ticks with repeats (ticks can't have hitnormals or additions, but repeats can). So it's fine in that regard.

Reply:
I had actually tested that it failed on the original rev, but then edited some things and forgot to re-test again. I've updated the comment above - should now fail before the change. There were not enough hitcircles to trigger a warning (didn't cross the 4 object threshold).